docs/source/working_with_aiida/caching.rst

```diff
@@ -65,88 +113,69 @@ This means that caching is enabled for ``TemplatereplacerCalculation`` and ``Str
+3. Run your calculation again. The node in questions should no longer be reused.
 
-Finally, workchains can create links not only to nodes which they create themselves, but also to nodes created by a calculation that they called, or even their ancestors. This is where caching becomes impossible. Consider the following example (using workfunctions for simplicity):
 
-.. code:: python
 
-    from aiida.engine import workfunction
-    from aiida.orm import Int
+.. _caching_limitations:
 
-    @workfunction
-    def select(a, b):
-        return b
+Limitations
+-----------
 
-    d = Int(1)
-    r1 = select(d, d)
-    r2 = select(Int(1), Int(1))
+1. The current implementation of caching for data nodes clones not only the graph representation of the reused node but also the underlying data in the database and file repository. 
+   This could be improved in order to reduce data duplication.
 
-The two ``select`` workfunctions have the same inputs as far as their hashes go. However, the first call uses the same input node twice, while the second one has two different inputs. If the second call should be cached from the first one, it is not clear which of the two input nodes should be returned.
+2. The caching mechanism for calculations *should* trigger only when the inputs and the calculation to be performed are exactly the same.  
+   Edge cases where this assumption might be violated include cases where the calculation parser is in a different python module than the calculation and the developer made changes without updating the version number of the plugin.
 
-While this example might seem contrived, the conclusion is valid more generally: Because workchains can return nodes from their history, they cannot be cached. Since even two equivalent workchains (with the same inputs) can have a different history, there is no way to deduce which links should be created on the new workchain without actually running it.
+3. The constraint that the shape of the provenance graph should be independent of whether caching is enabled or not, imposes limitations on the possible caching operations.
+   In order to reuse a cached node, we need to know not only its contents but also how the new node should be linked to its parents and children.
 
-Overall, this limitation is acceptable: The runtime of AiiDA workchains is usually dominated by time spent inside expensive calculations. Since these can be avoided with the caching mechanism, it still improves the runtime and required computer resources a lot.
+   * **Data nodes:** Making a copy of a data node should not change its links, so AiiDA just needs to link the new node to the direct ancestors of the old node.
```

Comment:
This is not what is currently happening and also not what should be happening I think.
```
#!/usr/bin/env runaiida
from aiida import orm
from aiida.engine import calcfunction


@calcfunction
def add(x, y):
    return x + y


x = orm.Int(1)
y = orm.Int(2)

result, node = add.run_get_node(x, y)

cached = orm.Int(result.value)
cached.store(use_cache=True)
print('Original', result.pk, result.get_incoming().all_nodes())
print('Cached', cached.pk, cached.get_incoming().all_nodes())
```
will print
```
('Original', 2996, [<CalcFunctionNode: uuid: 12f52293-3e3a-4604-9bef-4671f3a6bb82 (pk: 2995) (__main__.add)>])
('Cached', 2997, [])
```
Even though the `cloned` node has the same value as the output of the calcfunction, I don't think it should get the links of the parent. What does that even mean? Imagine we have many calculations that all produce an output integer with value 3, through different inputs and computations. When we now "cache" an `Int(3)`, which one of these output nodes do we take? To me this shows again why "caching" of `Data` nodes makes no sense and we should not talk about it. The only thing that is important is the "hashing" of data nodes

Reply:
I've removed the mention of data nodes in this list as I agree it is not helpful.